cloud-exec.md

```diff
@@ -0,0 +1,153 @@
+# Cloud execution
+
+- Enhancement Proposal PR: (leave this empty)
+- Contributors: dmpetrov
+
+# Summary
+
+Users need to run ML experiments in cloud instances or remote machines.
+DVC should support this out of the box to improve user experience and remove
+  the overhead of instrumenting these scenarios.
+
+# Motivation
+
+Today users need to spend time on instrumenting remote execution when there is
+  a lack of local resources for ML trainig (like GPU or RAM) or when it is a
+  company policy to train on dedicated hardware or on Kubernetes cluster.
+The existing way of executing training through [CML](cml.dev) (CI/CD) works
+  great for handing off ML models between team members but is not the best
+  approach in an experimentation phase since CI/CD creates an overhead on the
+  execution time and it creates some footprint/records in CI/CD that is
+  a bit overwhelming when the number of experiments is huge (10s, 100s or
+  1000s).
+This feature can save time for users and improve ML experimentation experience.
+
+## Scenarios
+
+1. **Run exp.** Allocate a cloud istance (with a specified configuration) and
+  execute an ML experiment `dvc exp run` (or `dvc repro`) on it.
+2. **Run stage.** Allocate a cloud istance and execute a particular stage of
+   an experiment on it.
+3. **Hot instance.** Execute on an existing running instance without instance
+   allocation.
+4. **Remote machine.** Execute in a remote machine (cloud or on-premise one).
+5. **Instance pool.** Execute on one of the existing running instances.
+6. **Exp queue** Execute a set of experiments (from exp queue for example) in
+   an instance pool or a single instance.
+7. **Web run.** Execute an experiment from SaaS\Web using one of the methods
+   from above and DVC machinery.
+
+## Optimizations
+
+Multiple cloud instance optimization technicques might be used to optimize
+  the usage of resources or execution time:
+1. **Spot instances.**
+2. **Transparent spot instance.** Recover the execution if a spot instance
+   was terminated. DVC checkpoints and pipeline stages should be used for
+   preserving the state.
+3. **Volumes.** Volumes can be attached and reused in instances to minimize
+   data cache synchronization time.
+4. **Shared volumes.** Separate cloud services (such as Multi-Attach EBS or
+   EFS) might be needed for sharing data cache between multiple instances.
```

Comment:
Should we separate compute resource types (1 and 2) and storage resource types (3 and 4)?

Reply:
Sure. But for now, the list of optimizations seems is very small to make separation or re-grouping. Also, some optimizations might be related - like re-attaching a volume when a spot instance fails.